### PR TITLE
Hilbert's Hotel orb now has a bunch of checks to prevent users pulling shenanigans

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

This item is a pain in the ass to test all the edge cases for. Please review carefully.

An item that doesn't do ANY sanity checks at all after `input`? This should have never passed review.

Hilbert's Hotel now does some basic level of sanity checking after the `input` to cut down on a number of shenanigans.

The following are all the new changes:
* Hilbert's cannot be activated via telekinesis. It must be activated by direct player attack.
* If the item isn't adjacent to the target after their input, it aborts.
* If the player attempting to enter the hotel is incapacitated after their input, it aborts.
* If the user no longer has the item in their possession after their input, it aborts even if there is a different target.
* If the user is also the target, it will attempt to drop the item normally. If the item fails to drop, it may be teleported to the hotel with the user. The item can fail to drop when it is in the user's possession but not in their hands, or when it has TRAIT_NODROP.
* When the item enters a hotel room created by itself, it will now recursively check for the first mob in its loc stack. It will gib this mob before teleporting away.
* This prevents the warp-whistle effect where a user could put the item in their backpack at INCREDIBLY low risk to themselves (0.135685% chance of anything bad happening) to enter a hotel room and have the orb warped away at random. They could then immediately leave for a free warp whistle effect.

## Why It's Good For The Game

I'm less likely to remove Hilbert's Hotel from the game a SECOND time, this time for being poorly coded and poorly designed.

You can still teleport yourself and your buddies to hotel rooms, but you can't warp whistle around, throw the orb then teleport to its new location, tag a friend and then run away and allow them to teleport into the orb 3 minutes later, so on, so forth.

"Improve, don't remove"

## Changelog
:cl:
balance: The Hilbert's orb Hotel can no longer be used via telekinesis, requires adjacency to teleport, will not teleport while incapacitated, requires you to have the item in your possession to teleport yourself and others and has new unforeseen consequences for attempting to transport the item into one of its own hotel rooms.
/:cl:

### CODE REVIEW

File: code/modules/ruins/spaceruin_code/hilbertshotel.dm [hunk 1]
```diff
@@ -334,6 +369,16 @@ GLOBAL_VAR_INIT(hhMysteryRoomNumber, rand(1, 999999))
...
 		return
+
+	// Prepare for...
+	var/mob/unforeseen_consequences = get_atom_on_turf(H, /mob)
+
+	// Turns out giving anyone who grabs a Hilbert's Hotel a free, complementary warp whistle is probably bad.
+	// Let's gib the last person to have selected a room number in it.
+	if(unforeseen_consequences)
+		to_chat(unforeseen_consequences, "<span class='warning'>\The [H] starts to resonate. Forcing it to enter itself induces a bluespace paradox, violently tearing your body apart.</span>")
+		unforeseen_consequences.gib()
```
Comment (Dennok) [inline code/modules/ruins/spaceruin_code/hilbertshotel.dm:380]: I dont fan of this, But why not first move `unforeseen_consequences` to targetturf (place where hilbertshotel land) before gibbing.

Comment (Timberpoes) [reply code/modules/ruins/spaceruin_code/hilbertshotel.dm:380]: That's the risk to using Hilbert's Hotel.

It's an item you can use to teleport away to some unknown location that nobody else can access without knowing the specific number you used between 1 and 16777216. You can then look through the peep hole as well as exit at any time you want.

Better be careful with how you use this power. Here's hoping you told someone what hotel room you were going to!

Comment (Dennok) [reply code/modules/ruins/spaceruin_code/hilbertshotel.dm:380]: A say about, move user to safeturf and next gib. to leave gibs on station level. not in hilbert hotel room. 
Or you want leave gibs in room?

Comment (Timberpoes) [reply code/modules/ruins/spaceruin_code/hilbertshotel.dm:380]: Gibbing them in the hotel room is the intent, yes. I hope they told their friends which room they're using if they try it.

**Comment by Acer202** [comment]
Am also not a fan of the gibbing chance. It's not part of HH lore and you literally justified it as "because screw you"

It's code-dive knowledge without any other tell. I'd implore you to add lore or get rid of that random chance, please!

**Comment by Timberpoes** [comment]
> 
> 
> Am also not a fan of the gibbing chance. It's not part of HH lore and it was (not you, but in general) literally justified it as "because screw you"
> 
> It's code-dive knowledge without any other tell. I'd implore you to add lore or get rid of that random chance, please!

The random chance wasn't added by me. It was there before.

I added the non-random chance.

**Comment by Acer202** [comment]
> The random chance wasn't added by me. It was there before.

Ah, I misunderstood, but it got me thinking:

Would it still gib you if, say, someone covered it in sticky tape and threw it on you when you tried to enter? That ever happening is extremely unlikely, but the possibility is funny.

**Comment by Timberpoes** [comment]
> 
> 
> > The random chance wasn't added by me. It was there before.
> 
> Ah, I misunderstood, but it got me thinking:
> 
> Would it still gib you if, say, someone covered it in sticky tape and threw it on you when you tried to enter? That ever happening is extremely unlikely, but the possibility is funny.

It shouldn't. It will fail if the orb is no longer in the possession of the person who initially used it you when you select a room number.

**Comment by Tharcoonvagh** [comment]
Just curious, do the hotel rooms have the `noteleport` flag?

**Comment by Dennok** [comment]
> Just curious, do the hotel rooms have the `noteleport` flag?

Yes. All transit space level has this flag, and hotel rooms appears in trsansit.

File: code/modules/ruins/spaceruin_code/hilbertshotel.dm (hunk 1 above)
Comment (cacogen) [reply code/modules/ruins/spaceruin_code/hilbertshotel.dm:380]: If you gib them in the hotel room anything they're carrying will be trapped there too. That could mean rare or unique items or traitor objectives being lost. Wouldn't it make more sense to gib them somewhere where these could conceivably be reached regardless of whether they've told people the room number (which they probably haven't).
